### ISSUE TITLE
Admin Consumer Pika Issue

### ISSUE THREAD

**Issue by hazmat345** [beer-garden/beer-garden#328]
A user is seeing an issue that looks very similar to this: https://github.com/pika/pika/issues/1219

Need to verify that what we're doing is still valid with pika > 1 from a thread-safety perspective.

**Comment by hazmat345** [comment]
One possibility - we should be using `add_callback_threadsafe`. There may be a potential issue when shutting down as the main thread is calling `stop_consuming` on the `RequestConsumer`s.

**Comment by hazmat345** [comment]
After looking into this a bit I'm pretty sure what we're doing was not valid, ever.

I was able to replicate the issue using 0.13.1 and 1.0.1. Every so often (but usually pretty quickly) I'd see a bunch of file descriptor write errors.

I believe the problem is due to how we're acking messages. To help me make sense of this I'm going to attempt to outline the way we process messages:

- pika invokes `RequestConsumer.on_message`
- `on_message` gives that message to the `RequestConsumer._on_message_callback`, which is `Plugin.process_[request|admin]_message`
- The plugin's `process_*_message` does some pre-processing and submits it to the relevant thread pool for *actual* processing
- The plugin returns the future that the thread pool gave it back to `RequestConsumer.on_message`
- `on_message` then uses `add_done_callback` to tack on `RequestConsumer.on_message_callback_complete` to the future so it runs when the processing finishes.
- `on_message_callback_complete` will check to see if the future has an exception.
  - If there is no exception it acks the message
  - If there *is* an exception on the future
    - If the exception is an instance of `DiscardMessageException` it nacks the message and does not requeue it
    - If the exception is an instance of `RepublishRequestException` it will construct an entirely new `BlockingConnection`, use that to publish a new message, and then ack the original message (This should be a very infrequent occurrence, but this seems kind of wild) 
    - If the exception is not an instance of either, or if there's ever an error acking a message, the `panic_event` is set and the consumer will self-destruct

So.

I think the issue is that wherever we're acking / nacking a message we're calling the method directly:
```python
self._channel.basic_ack(delivery_tag)
```

But it seems like tacking the `on_message_callback_complete` directly to the future causes that method to execute in the "wrong" context. From the Python docs:

> Added callables are called in the order that they were added and are always called in a thread belonging to the process that added them.

But the pika team has been very explicit - pika is not threadsafe and should not be called from other threads. The only exception to this is the `add_callback_threadsafe` method.

So instead of attaching `on_message_callback_complete` to the future directly, I attach this (using v1.0.1):

```python
 def on_message_callback_complete_intermediary(self, basic_deliver, future):
    self._connection.ioloop.add_callback_threadsafe(
        partial(self.on_message_callback_complete, basic_deliver, future)
    )
```

I haven't seen any errors yet, which is promising. But this is way too complicated to declare victory so quickly - I'm planning to investigate more.

**Comment by hazmat345** [comment]
This should have been closed by beer-garden/brewtils#153, not sure why it wasn't.
